Optimize IngestDocMetadata isAvailable

joegallo opened the pull request:
A good bit of the diff on the PR is fixing some src and test buglets around the code I'm changing -- it turns out the code and tests only appeared to work, but in reality they were both broken.

The actual work of this PR is that the `IngestDocMetadata` is formalized to have all its properties start with a leading underscore character (this was already true except for tests, and now it's true in tests, too), and then since we know that all metadata properties start with a leading underscore, we can shortcut the `isAvailable` check (which is a map `containsKey` call) in the case of a key with a leading character that is **not** an underscore and just return false. 

On the benchmark I'm running we call this a few hundred times per document, because we guard everything in `isAvailable` checks inside `CtxMap`, so even though this is the micro-est of optimizations it does actually matter in practice. My guess is that it gives better CPU cache locality because **we already have the key** but the properties map itself might be elsewhere in a worse cache line or whatever -- that is, it's faster this way because we don't have to run off to the heap in the vast majority of cases now (or at least that's the story I'm telling myself, it might not actually be true).

Here's a `rename` processor profile in the 'before' side, the purple bits are the `isAvailable` invocations (we really do call this a lot!):

<img width="1619" alt="Screenshot 2025-01-23 at 3 04 10 PM" src="https://github.com/user-attachments/assets/222568b7-2db8-43bb-93de-2fb2a4705149" />

joegallo committed:
Minor buglet: this throws a ConcurrentModificationException

You can't actually remove a key from a map while you're in the middle
of iterating through that same map. The only tests of this code
happened to pass because of the order of the keys in the tests (I
swear I am not making this up).

joegallo committed:
Micro improvement: don't look this up twice

joegallo committed:
Trivial cleanup Inline this map to save a variable

joegallo committed:
Fix a little test bug

joegallo committed:
Rename these keys to have leading underscores

joegallo committed:
Override isAvailable on IngestDocMetadata

and shortcut the map lookup in the almost-always-expected case of the
key being looked up on a CtxMap *not* having a leading underscore (and
therefore the whole isAvailable thing being an unnecessary expense).

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

joegallo committed:
Update docs/changelog/120753.yaml

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo commented on server/src/main/java/org/elasticsearch/ingest/IngestDocMetadata.java:
```diff
@@ -42,6 +42,20 @@ class IngestDocMetadata extends Metadata {
+
+    private static Map<String, FieldProperty<?>> validateLeadingUnderscores(final Map<String, FieldProperty<?>> properties) {
+        for (String key : properties.keySet()) {
+            assert key.charAt(0) == UNDERSCORE;
```
@masseyke `assert` made sense to me before I pulled this out into a method, now I'm thinking it should be an `IllegalArgumentException`, perhaps. What do you think?

joegallo replied:
I switched it in a683c54dd4f8ad18c2f8a3555d968585472e0846.

masseyke replied:
It looks like the constructor that lets you pass in arbitrary properties is only used by unit tests. It seems a little odd that we have this dangerous constructor and validation around it just for the sake of unit tests.

joegallo replied:
c090cbe62fb4b5c3e8f649e4f57c1e5a9bbcdf28 drops the bad constructor and fusses with a handful of tests that were using the more flexible version of things. I always hated `TestIngestCtxMetadata` so thanks for pushing on this, because now it's gone.

joegallo committed:
More test keys that need to start with underscores

joegallo committed:
Switch from an assert to a throws

masseyke commented on server/src/main/java/org/elasticsearch/script/CtxMap.java:
```diff
@@ -150,7 +151,7 @@ public Object remove(Object key) {
+        for (String key : new ArrayList<>(metadata.keySet())) { // copy the key set to get around the ConcurrentModificationException
             metadata.remove(key);
         }
         // TODO: this is just bogus, there isn't any case where metadata won't trip a failure above?
```
We could remove this TODO now

joegallo replied:
Ehhhhhhh, I rewrote the comment in ce216ff06d5b098cb2608b68baee83c6e65e6946, but I didn't remove it. `clear()` still doesn't work in the general case, for sure, and I'm not sure it works in any particularly practical case.

masseyke commented on server/src/main/java/org/elasticsearch/ingest/IngestDocMetadata.java:
```diff
@@ -100,4 +117,16 @@ private static void versionTypeValidator(String key, String value) {
+        assert key != null && key.isEmpty() == false;
+        // we can avoid a map lookup on most keys since we know that the only keys that are 'metadata keys' for an ingest document
+        // must be keys that start with an underscore
+        if (key.charAt(0) != UNDERSCORE) {
```
I'm amazed that this makes a difference at all (and I'm really curious why), but I've seen the charts!

masseyke reviewed (approved):
I would not have guessed that this would have made a measurable performance improvement, but it seems to. And it doesn't seem to cause any harm. LGTM

joegallo committed:
Merge branch 'main' into optimize-ingest-doc-metadata-is-available

joegallo committed:
Reword this comment

For example, this won't work for IngestCtxMap because the _version
isn't nullable, and so the metadata of an ingest document cannot be
clear()-ed.

joegallo committed:
Silence warnings from IntelliJ

joegallo committed:
Drop a constructor from IngestDocMetadata

We no longer allow arbitrary properties, the properties map is
hardcoded. TestIngestCtxMetadata is deleted entirely, and several
tests are rewritten to account for newer slightly-less-flexible (but
way easier to reason about!) IngestDocMetadata.

joegallo commented:
The last commit here is big enough that I'm re-requesting review -- I don't want to sneak anything by anybody.

joegallo committed:
These classes can be final
